python/pyiceberg/expressions/literals.py

```diff
@@ -309,23 +303,23 @@ def __init__(self, value: float):
         super().__init__(value, float)
 
     @singledispatchmethod
-    def to(self, type_var):
-        return None
+    def to(self, type_var: IcebergType) -> Literal:
+        raise TypeError(f"Cannot convert DoubleLiteral into {type_var}")
 
     @to.register(DoubleType)
-    def _(self, type_var: DoubleType) -> Literal[float]:
+    def _(self, _: DoubleType) -> Literal[float]:
         return self
 
     @to.register(FloatType)
-    def _(self, _: FloatType) -> Union[AboveMax, BelowMin, Literal[float]]:
+    def _(self, _: FloatType) -> Union[AboveMax, BelowMin, FloatLiteral]:
         if FloatType.max < self.value:
-            return AboveMax()
+            return AboveMax(FloatType.max, float)
```

Comment:
I see, it looks like this is why these are no longer singletons. What about `FloatAboveMax` and `IntegerAboveMax` instead? We will only have 2.

Reply:
Ah, I like that a lot!

Reply:
Added!